src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py

```diff
@@ -59,12 +70,95 @@ def _validation(target, data, evaluators, output_path, tracking_uri, evaluation_
+        # several columns and hence we cannot check the availability of columns
+        # without knowing target function semantics.
+        # Instead, here we will validate the columns, taken by target.
+        _validate_input_data_for_evaluator(target, None, df, is_target_fn=True)
+    else:
+        for evaluator_name, evaluator in evaluators.items():
+            _validate_input_data_for_evaluator(evaluator, evaluator_name, df)
+
+
+def _apply_target_to_data(
+        target: Callable,
+        data: str,
+        pf_client: PFClient,
+        initial_data: pd.DataFrame) -> Tuple[str, pd.DataFrame, List[str]]:
+    """
+    Apply the target function to the data set and save the data to temporary file.
+
+    :keyword target: The function to be applied to data.
+    :paramtype target: Callable
+    :keyword data: The path to input jsonl file.
+    :paramtype data: str
+    :keyword pf_client: The promptflow client to be used.
+    :paramtype pf_client: PFClient
+    :keyword initial_data: The data frame with the loaded data.
+    :paramtype initial_data: pd.DataFrame
+    :return: The tuple, containing path, data frame and the list of added columns.
+    :rtype: Tuple[str, pd.DataFrame, List[str]]
+    """
+    # We are manually creating the temporary directory for the flow
+    # because the way tempdir remove temporary directories will
+    # hang the debugger, because promptflow will keep flow directory.
+    saved_flow = f'flow_{uuid.uuid1()}'
+    os.makedirs(saved_flow)
+    save_function_as_flow(fun=target, target_dir=saved_flow, pf=pf_client)
+    run = pf_client.run(
+        flow=saved_flow,
+        data=data,
+        name=f'preprocess_{uuid.uuid1()}'
+    )
+    run = pf_client.stream(run)
+    # Delete temporary directory if we can.
+    try:
+        shutil.rmtree(saved_flow)
+    except BaseException:
+        # Exception means, we are running in debugger. In this case we can keep the
+        # directory.
+        pass
+    function_output = pf_client.runs.get_details(run, all_results=True)
+    # Remove input and output prefix
+    re_prefix = re.compile('(outputs[.])|(inputs[.])')
+    added_columns = {
+        re_prefix.sub('', col) for col in filter(
+            lambda x: x.startswith('outputs.'), function_output.columns)}
+
+    rename_dict = {col: re_prefix.sub('', col) for col in function_output.columns}
+    function_output.rename(columns=rename_dict, inplace=True)
+    drop_columns = set(function_output.columns) - added_columns
+    function_output.set_index(LINE_NUMBER, inplace=True)
+    function_output.sort_index(inplace=True)
+    function_output.reset_index(inplace=True, drop=False)
+    # function_output contains only input columns, taken by function,
+    # so we need to concatenate it to the input data frame.
+    function_output.drop(drop_columns, inplace=True, axis=1)
+    function_output = pd.concat([function_output, initial_data], axis=1)
+    new_data_name = f'{uuid.uuid1()}.jsonl'

```

Comment:
pf.run does accept dataframe. Please try that out and if that works we can avoid creating a new file.

Reply:
pf.run acepts string or [path like object](https://github.com/microsoft/promptflow/blob/4af2968d24279028d16e552a247f1073cdf1efed/src/promptflow-devkit/promptflow/_sdk/_pf_client.py#L218) If the data frame will be given, it will return an error as follows:
ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
because it [implicitly checks](https://github.com/microsoft/promptflow/blob/4af2968d24279028d16e552a247f1073cdf1efed/src/promptflow-devkit/promptflow/_sdk/_pf_client.py#L177) if data is None or empty.